lib/DBus/Client.hs

```diff
@@ -185,32 +228,132 @@ data ClientOptions t = ClientOptions
     , clientThreadRunner :: IO () -> IO ()
     }
 
-type Callback = (ReceivedMessage -> IO ())
-
 type FormattedMatchRule = String
 data SignalHandler = SignalHandler Unique FormattedMatchRule (IORef Bool) (Signal -> IO ())
 
+data Method = Method
+  { methodName :: MemberName
+  , inSignature :: Signature
+  , outSignature :: Signature
+  , methodHandler :: MethodCall -> IO Reply
+  }
+
+data Property = Property
+  { propertyName :: MemberName
+  , propertyType :: Type
+  , propertyGetter :: Maybe (IO Variant)
+  , propertySetter :: Maybe (Variant -> IO ())
+  }
+
 data Reply
     = ReplyReturn [Variant]
     | ReplyError ErrorName [Variant]
 
--- | Reply to a method call with a successful return, containing the given body.
-replyReturn :: [Variant] -> Reply
-replyReturn = ReplyReturn
-
--- | Reply to a method call with an error, containing the given error name and
--- body.
---
--- Typically, the first item of the error body is a string with a message
--- describing the error.
-replyError :: ErrorName -> [Variant] -> Reply
-replyError = ReplyError
-
-data Method = Method InterfaceName MemberName Signature Signature (MethodCall -> IO Reply)
+data Interface = Interface
+  { interfaceName :: InterfaceName
+  , interfaceMethods :: [Method]
+  , interfaceProperties :: [Property]
+  , interfaceSignals :: [I.Signal]
+  }
+
+defaultInterface :: Interface
+defaultInterface =
+  Interface { interfaceName = ""
+            , interfaceMethods = []
+            , interfaceProperties = []
+            , interfaceSignals = []
+            }
 
-type ObjectInfo = Map InterfaceName InterfaceInfo
-type InterfaceInfo = Map MemberName MethodInfo
-data MethodInfo = MethodInfo Signature Signature Callback
+data PathInfo = PathInfo
+  { _pathInterfaces :: [Interface]
+  , _pathChildren :: Map String PathInfo
+  }
+
+-- NOTE: This instance is needed to make modifyNothingHandler work, but it
+-- shouldn't really be used for much else.
+instance Eq PathInfo where
```

Comment:
Is it semantically correct that two path instances are only equal if both are null?
Given that PathInfo is exported, it is better to have correct instances, not just working for the narrow case.

Reply:
well you can't compare the interfaces because they cointain callable objects, and functions aren't comparable. Otherwise I would have provided a better implementation.

Reply:
Let's make a comment stating this clearly, then.